### DISCUSSION TITLE
How to programmatically insert or append Markdown content?

### DISCUSSION THREAD

**Discussion by forhappy** [mdx-editor/editor#229] [General]
## Use case

I want to implement an AI write feature on top of MDXEditor I'm wondering if we can support inserting or appending Markdown content at the current cursor and the editor will re-render after each AI write?

## Current Idea

I have implemented the floating toolbar and one button is "Continue with AI", when user clicks this Continue with AI button, I will call a backend API to generate Markdown content, the markdown content will be append to the current cursor. I have looked into the lexicaljs Autocomplete demo looks like that will be only useful for plain text.

## Help wanted
Can you help guide me on what API or plugin I can use to programmatically insert or append markdown content in MDXEditor?

**Comment by petyosi** [comment]
I might be missing something, but isn't the setMarkdown method going to work for you?

**Reply by forhappy** [reply]
Hey @petyosi thank you for replying. 

What I need is to implement the AI writing feature like Notion AI. The user can continue the writing with AI based on the current context or user prompts. The response from AI is streamed markdown content, I need to append the markdown to the current content.

Not sure setMarkdown can achieve this feature, my understanding is setmarkdown will reset the whole content and cause editor re-render.

My initial thought is to use the activeEditor reference and access the underlying lexical interfaces to implement the appending or inserting features.

**Reply by petyosi** [reply]
I see - if I understand correctly, you would like to display the streamed suggestion in a way that does not mutate the current actual state. Vim calls this virtual text or ghost text, I believe. 

IMO, this is a challenge, even on HTML level. I've done something like this by shimming a semi-transparent element with the new content, so that is visible but cannot participate in the selection or the cursor movement. I have not looked at what Lexical has in that direction, perhaps they solve it. If you can make it in Lexical, then it will work on its way out through the MDX conversion mechanism. 

Let me know what you've come up with. I do want to make such thing work down the road, but I'm not working on that right now.

**Reply by forhappy** [reply]
Thank you for your suggestion, will keep exploring the options.

**Comment by forhappy** [comment]
I just raised PR https://github.com/mdx-editor/editor/pull/256 to implement this feature :)
